proposals/lip-define_new_transaction_schema.md

```diff
@@ -0,0 +1,190 @@
+Created: <YYYY-MM-DD>
+Updated: <YYYY-MM-DD>
+```
+
+## Abstract
+
+This LIP defines a new schema to be used to serialize transactions. The main change is to make module and command identifiers to be of type bytes. This LIP also updates the terminology used for transaction and transaction properties. 
+
+
+## Copyright
+
+This LIP is licensed under the [Creative Commons Zero 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/).
+
+
+## Motivation
+
+The Lisk protocol handles identifiers for transactions, modules, commands, and many more. The type of those identifiers is however not fully consistent, as some are of type `uint32` (like module ID and chain ID) and others of type `bytes` (like transaction ID and block ID). Moreover, all identifiers used in the new state model to compute the store keys must be first converted to type `bytes`. Unifying all identifier types to be of type `bytes` simplifies their handling and avoids unnecessary type conversion.
+
+
+## Rationale
+
+
+### Unifying Identifier Type
+
+The type identifiers in the Lisk protocol are not fully consistent, as some are of type `uint32` and others of type `bytes`. This means that special care is required to use the proper type whenever the identifier is used. Further, identifiers are often used to compute the store keys in the state tree and for this purpose must always be of type bytes. Hence, identifiers of `uint32` type need to be converted to fulfill this purpose. This implies that the implementation very often converts those identifiers from their integer form (as schema entries) to the corresponding bytes (as store key). The new transaction schema introduced in this LIP sets the module ID and command ID to type bytes. 
+
+Defining identifiers as type `bytes` also requires fixing the length of the identifier. This was not possible when using identifiers of type `uint32`, as the full 4 bytes of the maximal range always had to be assumed when using the identifier in the state tree.  
+
+
+### New Property Names
+
+All properties in the proposed transaction schema are equivalent to the ones defined in LIP 0028. The only changes are the renaming of `assetID` to `commandID` and of `asset` to `params`. As was described in LIP "Update Lisk SDK modular blockchain architecture".
+
+
+## Specification
+
+The transaction schema defined in [LIP 0028](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0028.md) is superseded by the one defined below. 
+
+The `params` property must follow the schema corresponding to the `moduleID`, `commandID` pair defined in the corresponding module.
+
+All transaction procedures - serialization, deserialization, signature calculation, signature validation and transaction ID - follow the same specifications already defined in [LIP 0028](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0028.md#serialization). The resulting serialization or signatures are however different when the proposed transaction schema is used.
```

Comment:
I think this is actually not true for the following reasons:
- We do not allow optional properties in schemas any more due to LIP 64. Therefore, we need a schema both for signed and unsigned transactions.
- We now use message tags for signing transactions as defined in LIP 37.

I would suggest to define these functions here using Python-style pseudocode with type-hinting. This way we also have a clean interface to use in other places.

Reply:
As far as I see these would be the functions:
```
encode(assetSchema: LiskJSONSchema, trs: Object) -> bytes:

decode(assetSchema: LiskJSONSchema, transactionData: bytes) -> Object:

signTransaction(assetSchema: LiskJSONSchema, trs: Object) -> SignatureEd25519:

verifyTransactionSignature(assetSchema: LiskJSONSchema, trs: Object) -> bool:
```

Reply:
@ricott1 Are you assuming there is a property `trs.schema` and `trs.assetSchema` or where should the schema come from? Apart from the interface to the state store and constants, I would try to avoid introducing global variables which are not defined anywhere.

Reply:
As mentioned [here](https://github.com/LiskHQ/lips/pull/155#discussion_r937632956) I would still suggest to introduce a stateless signature validation function so that this LIP defines serialization/deserialization, signing/signature validation for transactions fully. Then this functionality is encapsulated in one place and could be updated (e.g., changing the message tag, call to the underlying crypto library) without changing many LIPs in different places.

Reply:
To me it could work in a way similar to how we call commands or methods (for instance in the interop LIPs): we say, `assetSchema = JSON schema corresponding to (module, command) tuple`.

Reply:
I would agree with @ricott1 for the encode function, but for the decode you can not find the paramsSchema in the same way (since you don't have access to module,command). So I include `paramsSchema` as an argument

Reply:
also for decoding, you first decode the root schema, and then get the paramsSchema from the module and command

Reply:
Looks good now, can be resolved.